internal/controller/operator/factory/vmagent/vmagent.go

```diff
@@ -275,88 +184,90 @@ func createOrUpdateShardedDeploy(ctx context.Context, rclient client.Client, cr,
-			if prevDeploy != nil {
-				prevShardedObject = prevDeploy.DeepCopyObject()
-				addShardSettingsToVMAgent(shardNum, shardsCount, prevShardedObject)
+	updateShard := func(num *int) {
+		var rv returnValue
+		defer func() {
+			rtCh <- &rv
+			wg.Done()
+		}()
+
+		if cr.Spec.PodDisruptionBudget != nil && !cr.Spec.DaemonSetMode {
+			pdb := build.PodDisruptionBudgetSharded(cr, cr.Spec.PodDisruptionBudget, num)
+			var prevPDB *policyv1.PodDisruptionBudget
+			if prevCR != nil && prevCR.Spec.PodDisruptionBudget != nil {
+				prevPDB = build.PodDisruptionBudgetSharded(prevCR, prevCR.Spec.PodDisruptionBudget, num)
 			}
-			placeholders := map[string]string{shardNumPlaceholder: strconv.Itoa(shardNum)}
-
-			switch shardedDeploy := shardedDeploy.(type) {
-			case *appsv1.Deployment:
-				var prevDeploy *appsv1.Deployment
-				var err error
-				shardedDeploy, err = k8stools.RenderPlaceholders(shardedDeploy, placeholders)
-				if err != nil {
-					rv.err = fmt.Errorf("cannot fill placeholders for deployment sharded vmagent(%d): %w", shardNum, err)
-					return
-				}
-				if prevShardedObject != nil {
-					// prev object could be deployment due to switching from statefulmode
-					prevObjApp, ok := prevShardedObject.(*appsv1.Deployment)
-					if ok {
-						prevDeploy = prevObjApp
-						prevDeploy, err = k8stools.RenderPlaceholders(prevDeploy, placeholders)
-						if err != nil {
-							rv.err = fmt.Errorf("cannot fill placeholders for prev deployment sharded vmagent(%d): %w", shardNum, err)
-							return
-						}
-					}
-				}
+			if err := reconcile.PDB(ctx, rclient, pdb, prevPDB); err != nil {
+				rv.err = err
+				return
+			}
+		}
 
-				if err := reconcile.Deployment(ctx, rclient, shardedDeploy, prevDeploy, false); err != nil {
-					rv.err = fmt.Errorf("cannot reconcile deployment for sharded vmagent(%d): %w", shardNum, err)
+		switch newAppTpl := newAppTpl.(type) {
+		case *appsv1.Deployment:
+			newApp, err := build.RenderShard(newAppTpl, nil, num)
+			if err != nil {
+				rv.err = err
+				return
+			}
+			patchShardContainers(newApp.Spec.Template.Spec.Containers, num, shardCount)
+			var prevApp *appsv1.Deployment
+			if prevAppTpl != nil {
+				if prevApp, err = build.RenderShard(prevAppTpl.(*appsv1.Deployment), nil, num); err != nil {
+					rv.err = err
 					return
 				}
-				rv.deploymentName = shardedDeploy.Name
-			case *appsv1.StatefulSet:
-				var prevSts *appsv1.StatefulSet
-				var err error
-				shardedDeploy, err = k8stools.RenderPlaceholders(shardedDeploy, placeholders)
+				patchShardContainers(prevApp.Spec.Template.Spec.Containers, num, shardCount)
+			}
+			if err = reconcile.Deployment(ctx, rclient, newApp, prevApp, false); err != nil {
+				rv.err = fmt.Errorf("cannot reconcile deployment for sharded vmagent(%d): %w", num, err)
+				return
+			}
+			rv.deploymentName = newApp.Name
+		case *appsv1.StatefulSet:
+			newApp, err := build.RenderShard(newAppTpl, nil, num)
+			if err != nil {
+				rv.err = err
```

Comment:
```suggestion
				rv.err = fmt.Errorf("cannot render new Statefulset shard: %w", err)
```
nitpick: let's wrap errors earlier so that we can spot a precise point where it happened